src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1792,7 +1792,7 @@ bool Compiler::fgVNBasedIntrinsicExpansionForCall_ReadUtf8(BasicBlock** pBlock,
 //
 PhaseStatus Compiler::fgLateCastExpansion()
 {
-    if (!doesMethodHaveExpandableCasts() || opts.OptimizationDisabled())
+    if (!doesMethodHaveExpandableCasts() || opts.OptimizationDisabled() || lvaHaveManyLocals())
```

Comment:
I'd generally be opposed to this check, which makes it quite easy to hit performance cliffs (change your code slightly -> hit more locals -> casts are no longer expanded globally in the function). What was the motivation for it?

Reply:
@jakobbotsch repeats the existing check in importer, although, I should probably follow the pattern and only reject non-local objects (before tree split) - https://github.com/dotnet/runtime/blob/f582c0b66b2b91b626ccce62ea0e37496446e79f/src/coreclr/jit/importer.cpp#L5459-L5462

This was responsible for size regressions in some tests

Reply:
Also, worth experimenting how increased threshold of locals we can try affects diffs, TP but it's irrelevant to this PR 🙂

Reply:
In the importer this check corresponds much more directly to user locals. But this late the correspondence is much less direct since the JIT can have created a number of new locals in many more places (e.g. promotion, call args lowering, CSE, ...). I wonder if instead we can add some bailout for pathological cases where we've expanded tons of casts already, or maybe start reusing locals if we've already created a large amount.

Reply:
`lvaHaveManyLocals` just looks at `lvaCount`, so even w/o recycling we can periodically keep track of how many useless locals there are and subtract that part off. Or just use the number of tracked locals instead.

Renumbering or reusing locals seems trickier to pull off.

Reply:
@jakobbotsch I've changed to logic in https://github.com/dotnet/runtime/pull/98273/commits/935f6e21de441cb783229d72e798c292100f24db a bit to match importer's behavior, although, I see your point - do you mind if I think about the best way to handle it separately? I'd love to get rid of the importer's expansion as soon as possible so then I can focus on polishing the late one.

Reply:
Sure.